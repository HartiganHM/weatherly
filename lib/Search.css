.Search {
  display: flex;
  height: 4rem;
}

input, button {
  border: 0.02rem solid ghostwhite;
  font-size: 1.8rem;
}

input {
  flex: 2;
  padding-left: 1rem;
}