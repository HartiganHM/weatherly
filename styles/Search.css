.Search {
  display: flex;
  flex-direction: column;
}

.citySuggestion {
  background-color: rgb(37, 39, 53);
  color: #FFF;
  padding: 1.5rem;
  width: 100%;
  z-index: 5;
}

.citySuggestion:hover {
  background-color: rgb(67, 70, 94);
}

.Inputs {
  display: flex;
  height: 4rem;
}

.citySuggestion, .searchField, .searchButton {
  border: 0.02rem solid ghostwhite;
  font-size: 1.4rem;
  outline: none;
}

.searchButton {
  border-radius: 0 2rem 2rem 0;
}

.searchField {
  border-radius: 2rem 0 0 2rem;
  flex: 2;
  padding-left: 1rem;
}

.suggestBox {
  display: flex;
  flex-direction: column;
  height: 0;
  list-style: none;
  margin: auto;
  margin-bottom: 2rem;
  padding: 0;
}