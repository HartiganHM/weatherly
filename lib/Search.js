import React, { Component } from "react";
import Trie from "@hartiganhm/linked-list";
import CitiesList from "./AutoCompleteData";
import "../styles/Search.css";

export default class Search extends Component {
  constructor() {
    super();
    this.trie = new Trie("");
    this.trie.populate(CitiesList.data);

    this.autoComplete = this.autoComplete.bind(this);
    this.enterHit = this.enterHit.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleClickChange = this.handleClickChange.bind(this);
    this.reFocus = this.reFocus.bind(this);

    this.suggestions = [];
    this.state = {
      value: "",
      location: [],
      index: -1
    };
  }

  handleInputChange(event) {
    this.setState({
      value: event.target.value
    });

    if (event.target.value.length) {
      this.suggestions = this.trie.suggest(event.target.value);
    } else {
      this.suggestions = [];
    }
  }

  autoComplete(event) {
    this.setLocation(event.target.innerText.split(", "));
  }

  enterHit(event) {
    if (this.state.value.length && event.key === "Enter") {
      this.setLocation((location = this.state.value.split(", ")));
    }
  }

  handleClickChange(event) {
    this.setLocation((location = this.state.value.split(", ")));
  }

  setLocation(location) {
    this.setState({
      location: location,
      value: ""
    });
    this.suggestions = [];
    localStorage.setItem("location", JSON.stringify(location));
    this.props.fetch(location);
  }

  reFocus(e) {
    let indexVal = this.state.index;
    
    if (e.key === "ArrowDown") {
      this.setState({
        index: Math.min(indexVal + 1, 5)
      });
    } else if (e.key === "ArrowUp") {
      this.setState({
        index: Math.max(indexVal - 1, 0)
      });
    }
  }

  render() {
    return (
      <div className="Search">
        <div className="Inputs">
          <input
            className="searchField"
            type="text"
            value={this.state.value}
            onChange={this.handleInputChange}
            onKeyDown={this.reFocus}
            placeholder="Search by city or zip for Weatherly Forecast"
            autoFocus
          />
          <button className="searchButton" onClick={this.handleClickChange}>
            Search
          </button>
          {!this.props.greeting && <p className="branding">Weatherly</p>}
        </div>
        <ul className="suggestBox">
          {this.suggestions.map((suggestion, index) => {
            let classes = "citySuggestion";

            if (index === this.state.index) {
              classes = `${classes} highlight`;
              this.state.value = this.suggestions[index].toUpperCase();
            }

            if (index < 5) {
              return (
                <li className={classes} onClick={this.autoComplete} key={index}>
                  {suggestion.toUpperCase()}
                </li>
              );
            }
          })}
        </ul>
      </div>
    );
  }
}
