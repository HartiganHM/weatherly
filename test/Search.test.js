import React from 'react';
import { shallow, mount } from 'enzyme';
import Search from '../lib/Search';


describe('Search', () => {
  it('changing input should change state', () => {
    const search = shallow(<Search filterFunction={ ()=>{} }/>)
    const input = search.find('input');

    const mockEvent = {
      target: {
        value: 'Solar'
      }
    }

    expect( search.state('inputVal') ).toEqual('')

    input.simulate('change', mockEvent);

    expect( search.state('inputVal') ).toEqual('Solar')

  })
});
