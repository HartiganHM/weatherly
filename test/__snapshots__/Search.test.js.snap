// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should have a className search 1`] = `[Function]`;
